### PR TITLE
Better handling of remember me token when login across devices is disabled

### LINKED ISSUE #9794: UX weakness with "remember me" leading to unnecessary forgotten logins
There are a couple of issues with the "remember me" function that lead to it giving the appearance of being unreliable. This probably has flow-on issues such as resistance to rolling out MFA as it relies on this function for a good UX.

In my view, the first scenario has the higher impact, but that's a bit subjective.

## Scenario 1
Steps:

 * Log in from Safari, tick remember me
 * Close and restart Safari, verify that login has been remembered
 * Log in from Chrome, do not tick remember me
 * Close and restart Safari

Expected:
 * Login has been remembered
 
Actual
 * Login on Safari has been forgotten

### Explanation

I believe that when the "remember me" box is not checked, it is clearing out any ALC token from all devices. It would make sense to clear out the ALC from _this_ device, but not other devices.

From my perspective, this is a bug, without any security implications.

## Scenario 2
Steps:

 * Log in from Safari, tick remember me
 * Close and restart Safari, verify that login has been remembered
 * Log in from Chrome, tick remember me
 * Close and restart Safari, verify that login has been remembered
 * Explicitly log out from Chrome
 * Reload Safari, verify that it hasn't been logged out
 * Close and restart Safari

Expected:
 * Login has been remembered
 
Actual
 * Login on Safari has been forgotten

### Explanation

The "log out" functionality is hard-coded to perform "log out from all devices". This has some security benefit but doesn't make a lot of sense as a default.

From my perspective, this is a bug, but it _does_ have some security implications. Ideally, the log out action (at least in the CMS) would pop up a box with a "log out from all devices" checkbox, defaulting to unchecked.

### ACs
- [ ] A technical risk assessment is performed assuming the below is implemented
- [x] Neglecting to check the remember me checkbox in one browser does not remove settings for other browsers
- [ ] By default, logging out of one browser session does not invalidate other browser sessions (currently managed in the `true/false` configuration value)
- [x] Review existing documentation and any changes required based on this fix
- [x] Changelog notes this as a bug, not a feature change

### Update on ACs
- We choose to go with a 'do mostly nothing' approach so in the end there was no need for a technical risk assessment.  We choose not to change the default behaviour of logging out of one browser invalidating other sessions (that has check the remember me checkbox)

### Notes
- This fix should be tested with the silverstripe/session-manager module - we'll cover that in a separate issues

# Pull request
- [x] #9895

### PR DESCRIPTION
**Pull request by maxime-rainville**
The specific bug addressed by this fix is that when you logged out from a non-persistent session (you did NOT check the remember checkbox) and `login across devices` is disabled, you will not the less invalidate all remember me token for this user.

# Parent issue
Fixes #9794

### CODE REVIEW

**Comment by maxime-rainville** [comment]
Still need to review a bit of doc

**Comment by maxime-rainville** [comment]
Reviewed the relevant doc for this PR. It looks OK and there isn't any obvious improvement I would make at this stage.

https://github.com/silverstripe/silverstripe-framework/blob/21129b1624b21c95d51e1cb79e394d1030baf4a5/docs/en/02_Developer_Guides/09_Security/00_Member.md#L137-L142

File: src/Security/RememberLoginHash.php [hunk 1]
```diff
@@ -177,8 +177,17 @@ public function renew()
     public static function clear(Member $member, $alcDevice = null)
     {
         if (!$member->exists()) {
+            // If we don't have a valid user, we can't clear theic Remember me token
```
Comment (maxime-rainville) [inline src/Security/RememberLoginHash.php:180]:
```suggestion
            // If we don't have a valid user, we can't clear any "Remember me" tokens
```

File: src/Security/RememberLoginHash.php [hunk 2]
```diff
@@ -177,8 +177,17 @@ public function renew()
     public static function clear(Member $member, $alcDevice = null)
     {
         if (!$member->exists()) {
+            // If we don't have a valid user, we can't clear any "Remember me" tokens
             return;
         }
+
+        $logoutAcrossDevices = static::config()->logout_across_devices;
```
Comment (emteknetnz) [inline src/Security/RememberLoginHash.php:184]: This code is fine and works correctly, though read together with the block below with the existing logic, it's quite hard to follow it all (it's basically that existing code is hard to read)

Here's a proposed rewrite that makes everything a lot easier to read

```
        if (static::config()->logout_across_devices) {
            RememberLoginHash::get()->filter([
                'MemberID' => $member->ID
            ])->removeAll();
            return;
        }

        if ($alcDevice) {
            RememberLoginHash::get()->filter([
                'MemberID' => $member->ID,
                'DeviceID' => $alcDevice
            ])->removeAll();
        }
```

File: tests/php/Security/RememberLoginHashTest.yml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+'SilverStripe\Security\Member':
```
Comment (emteknetnz) [inline tests/php/Security/RememberLoginHashTest.yml:1]: I'm honestly surprised you used yml fixtures given all your previous DataObject hydration work :-)

**Review by emteknetnz — APPROVED** [review]
Code looks good.

Recommended rewrite to clarify the code (mostly the existing code), though also fine with merging as is

File: tests/php/Security/RememberLoginHashTest.php [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+namespace SilverStripe\Security\Tests;
+
+use SilverStripe\Dev\SapphireTest;
+use SilverStripe\Security\Member;
+use SilverStripe\Security\RememberLoginHash;
+
+/**
+ * @author Ingo Schommer
+ */
```
Comment (maxime-rainville) [inline tests/php/Security/RememberLoginHashTest.php:11]:
```suggestion
```

File: tests/php/Security/RememberLoginHashTest.yml (hunk 3 above)
Comment (maxime-rainville) [reply tests/php/Security/RememberLoginHashTest.yml:1]: Old habbits

File: src/Security/RememberLoginHash.php (hunk 2 above)
Comment (maxime-rainville) [reply src/Security/RememberLoginHash.php]: Tweaked the logic as suggested.

File: src/Security/RememberLoginHash.php [hunk 5]
```diff
@@ -172,19 +172,25 @@ public function renew()
...
+            self::get()->filter([
+                'MemberID' => $member->ID,
+                'DeviceID' => $alcDevice
+            ])->removeAll();
         }
-        RememberLoginHash::get()
-            ->filter($filter)
-            ->removeAll();
+
+        // We've logged in without checking the "Remember me" checkbox and `logout_across_devices` is disable ...
```
Comment (emteknetnz) [inline src/Security/RememberLoginHash.php:193]: This comment is pretty random and doesn't really make sense in this context.  It almost needs to be wrapped in an else statement.  Could also just delete it.

**Review by emteknetnz — APPROVED** [review]
Random comment that should probably be removed.  OK to merge as is though.

File: src/Security/RememberLoginHash.php [hunk 6]
```diff
@@ -172,19 +172,25 @@ public function renew()
...
+        // so we don't have any RememberLoginHash to clear
```
Comment (maxime-rainville) [inline src/Security/RememberLoginHash.php:194]:
```suggestion
```

**Comment by maxime-rainville** [comment]
Merge on green
